File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:

        brackets = {'{':'}','[':']','(':')'}
        stack = []

        for b in s:

            if b in brackets.keys():
                stack.append(b)

            else:
                if len(stack) == 0 or brackets[stack[-1]] != b:
                    return False

                stack.pop()

        if len(stack) == 0:
            return True

        else:
            return False
        

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.ts
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates: number[], target: number): number[][] {

        candidates.sort((a,b) => a-b)

        const res:number[][] = [];
        

        function backtrack(i:number, com: number[], sum: number) {

            if (sum === 0) {
                res.push(com)
                return
            }

            else if (sum < 0 || i >= candidates.length) {
                return
            }

            let val = candidates[i]
            i++
            com.push(val)
            backtrack(i, [...com], sum - val)

            com.pop()
            while (i < candidates.length && candidates[i] === val) {
                i++
            }
            backtrack(i, [...com], sum)
        }

        backtrack(0,[], target)
        return res

    }
}
